Add SUBSCRIBE_UPDATE (#385)
moq-wg/moq-transport: draft-ietf-moq-transport.md

|-------|-----------------------------------------------------|
| 0x9   | UNANNOUNCE  ({{message-unannounce}})                |
|-------|-----------------------------------------------------|
| 0xA   | SUBSCRIBE_UPDATE ({{message-subscribe-update}})     |

> |-------|-----------------------------------------------------|

Reviewer: I feel Unsubscribe makes more sense if one wants to end the subscription.
We don;t need a new message to update an existing subscription. A subscriber can update an existing subscription by sending `SUBSCRIBE` message with SubscribeId that matches an existing subscription.

|-------|-----------------------------------------------------|
| 0x9   | UNANNOUNCE  ({{message-unannounce}})                |
|-------|-----------------------------------------------------|
| 0xA   | SUBSCRIBE_UPDATE ({{message-subscribe-update}})     |

> |-------|-----------------------------------------------------|

Reviewer 2: I think we need both the UNSUBSCRIBE and UPDATE ( or whatever names we decide to call them ).

  the receiver MUST close the connection with a Duplicate Track Alias error
  ({{session-termination}}).

## SUBSCRIBE_UPDATE {#message-subscribe-update}

A subscriber issues a `SUBSCRIBE_UPDATE` message to a publisher to request a
modification to the end of an existing subscription.  One common type of
modification is unsubscription.

> If an update cannot be completed by the relay, it replies with a SUBSCRIBE_ERROR

Reviewer 3: But what about SUBSCRIBE_OK? We should send it too, especially if the end groups use relative offsets.

Reviewer 3: Also what if there are multiple SUBSCRIBE_UPDATES? Do the SUBSCRIBE_ERRORs have to be returned in order? With no OK, it's impossible to differentiate if the 1st or 2nd UPDATE failed.

> with error code 'Update Failed' and the subscription is unchanged.

Reviewer 3: I don't like the subscription being unchanged. It causes a partial failure which is just really difficult to support.

A SUBSCRIBE_UPDATE MUST NOT fail if the specified EndGroup and EndObject are
earlier than the current values for the subscription.

The subscription is not fully terminated until a SUBSCRIBE_RST or SUBSCRIBE_FIN
are received.

The format of `SUBSCRIBE_UPDATE` is as follows:

~~~

> SUBSCRIBE_UPDATE Message {

Reviewer 3: Can we also update the start group/object? It would be super useful for relays to deal with:

viewer1: SUBSCRIBE start=-2
(pause)
viewer2: SUBSCRIBE start=-3

Currently, the relay would issue to origin:
SUBSCRIBE start=-2
(wait for SUBSCRIBE_OK -2=65)
SUBSCRIBE start=64 end=65

Which is complicated and incurs delay. But if we allow updating start ranges:
SUBSCRIBE start=-2
SUBSCRIBE_UPDATE start=-3

Reviewer 4: Updating the start point can mess with object mapping to streams, especially moving backwards.  In your example, if the track is stream-per-track, the publisher may open a stream and being publishing group=65.  In the current design, where subscription ID appears in objects, the second subscribe for group=64 will come on its own new stream.  If we allow updating the subscription start point backwards to 64, what should the publisher do? Should it close the open stream and open a new one, beginning at group=64, and resend objects from group=65?

Author: I do think updating the start is potentially interesting and useful.

Updates of the start should not be able to move the start point of the subscription earlier.  As you said, it's just too complex.  For similar reasons, I think updating the end to be farther in the future is error prone as well, though there may be use cases for that.

Reviewer 2: +1 on start can only be moved forward, end can only be moved backwards

Reviewer 2: As a side note where I don't really care ... I'm not sure the use case to move the start forward

  Subscribe ID (i),
  EndGroup (Location),

>   EndObject (Location),

Reviewer: This needs to be optional and only set if the subscription needs to be ended.

Reviewer 4: In the context of this PR and the current draft, where the only thing you can change is the end, this message is OK.  However, if we add more parameters to subscribe (eg #381), then we need a way to update some of the parameters and not others.

We should also clarify what happens if you issue the SUBSCRIBE twice for the same ID - I expect in this PR as is, we would make that an error?

  the receiver MUST close the connection with a Duplicate Track Alias error
  ({{session-termination}}).

## SUBSCRIBE_UPDATE {#message-subscribe-update}

> A subscriber issues a `SUBSCRIBE_UPDATE` message to a publisher requesting a

Author: ```suggestion
A subscriber issues a `SUBSCRIBE_UPDATE` message to a publisher to request a
```

modification to the end of an existing subscription.  One common type of
modification is unsubscription.

> If an update cannot be completed by the relay, it replies with a 'SUBSCRIBE_ERROR'

Author: ```suggestion
If an update cannot be completed by the relay, it replies with a SUBSCRIBE_ERROR
```

  the receiver MUST close the connection with a Duplicate Track Alias error
  ({{session-termination}}).

## SUBSCRIBE_UPDATE {#message-subscribe-update}

A subscriber issues a `SUBSCRIBE_UPDATE` message to a publisher to request a
modification to the end of an existing subscription.  One common type of
modification is unsubscription.

If an update cannot be completed by the relay, it replies with a SUBSCRIBE_ERROR
with error code 'Update Failed' and the subscription is unchanged.

> A SUBSCRIBE_UPDATE MUST NOT fail if the specified EndGroup and EndObject have
> already been received.

Author: ```suggestion
A SUBSCRIBE_UPDATE MUST NOT fail if the specified EndGroup and EndObject are
earlier than the current values.
```

> A SUBSCRIBE_UPDATE MUST NOT fail if the specified EndGroup and EndObject are
> earlier than the current values.

Author: ```suggestion
A SUBSCRIBE_UPDATE MUST NOT fail if the specified EndGroup and EndObject are
earlier than the current values for the subscription.
```

  the receiver MUST close the connection with a Duplicate Track Alias error
  ({{session-termination}}).

## SUBSCRIBE_UPDATE {#message-subscribe-update}

A subscriber issues a `SUBSCRIBE_UPDATE` message to a publisher to request a

> modification to the end of an existing subscription.  One common type of

Reviewer 2: I think the "end of existing" should say the "end of the subscription range".

> modification is unsubscription.

Reviewer 2: I don't think this is right. The update should just be to change the metadata about the subscribe, not change the state of the subscribe. We should use some other message to cause the SUSBCRIBE to end and remove all the state about the SUBSCRIBE.

> If an update cannot be completed by the relay, it replies with a SUBSCRIBE_ERROR

Reviewer 2: I think this should be an UPDATE_OK/ERROR. Don't care if it is two message or not. But should be clear that if some part of the UPDATE can not be applied, then no part of the update is applied, subscribe stays with same meta data as before, and error returned. So an update is all or nothing.

with error code 'Update Failed' and the subscription is unchanged.
A SUBSCRIBE_UPDATE MUST NOT fail if the specified EndGroup and EndObject are
earlier than the current values for the subscription.

The subscription is not fully terminated until a SUBSCRIBE_RST or SUBSCRIBE_FIN

> are received.

Reviewer 2: Uh, I think this needs to be UNSUBSCRIBE or whatever the  SUBSCRIBE CLOSE/STOP/FIN/RST/END is called

The format of `SUBSCRIBE_UPDATE` is as follows:

~~~
SUBSCRIBE_UPDATE Message {
  Subscribe ID (i),

>   EndGroup (Location),

Reviewer 2: This is not a location this shoudl be an optional var int

>   EndObject (Location),

Reviewer 2: This should be var int not Location

}
~~~
{: #moq-transport-subscribe-update-format title="MOQT SUBSCRIBE_UPDATE Message"}

* Subscribe ID: Subscription Identifer as defined in {{message-subscribe-req}}.

* EndGroup: The last Group requested in the subscription, inclusive.  EndGroup's

> Mode is None for an open-ended subscription.

Reviewer 2: I would prefer to make the end group / object optional in the update instead of having a None value. I think it makes a better patter for how to add future things to the update.
